src/treelearner/monotone_constraints.hpp

```diff
@@ -78,48 +365,148 @@ class LeafConstraintsBase {
+      FeatureMinOrMaxConstraints &constraints_to_be_updated =
+          constraints[feature_for_constraint].FeatureMinConstraintsToBeUpdated()
+              ? constraints[feature_for_constraint].GetMinConstraints()
+              : constraints[feature_for_constraint].GetMaxConstraints();
+
+      constraints_to_be_updated.Reset(
+          constraints[feature_for_constraint].FeatureMinConstraintsToBeUpdated()
+              ? -std::numeric_limits<double>::max()
+              : std::numeric_limits<double>::max());
+
+      std::vector<int> features_of_splits_going_up_from_original_leaf =
+          std::vector<int>();
+      std::vector<uint32_t> thresholds_of_splits_going_up_from_original_leaf =
+          std::vector<uint32_t>();
+      std::vector<bool> was_original_leaf_right_child_of_split =
+          std::vector<bool>();
+      constraints_->GoUpToFindConstrainingLeaves(
+          feature_for_constraint, leaf_idx,
+          &features_of_splits_going_up_from_original_leaf,
+          &thresholds_of_splits_going_up_from_original_leaf,
+          &was_original_leaf_right_child_of_split, &constraints_to_be_updated,
+          constraints[feature_for_constraint]
+              .FeatureMinConstraintsToBeUpdated(),
+          0, it_end, it_end);
+      constraints[feature_for_constraint].ResetUpdates();
+    }
+  }
+
+  // for each feature, an array of constraints needs to be stored
+  explicit AdvancedConstraintEntry(unsigned int num_features) {
+    constraints.resize(num_features);
+  }
+
+  void Reset() final {
+    for (unsigned int i = 0; i < constraints.size(); i++) {
+      constraints[i].Reset();
+    }
+  }
+
+  void UpdateMin(double new_min) final {
+    for (unsigned int i = 0; i < constraints.size(); i++) {
+      constraints[i].UpdateMin(new_min, false);
+    }
+  }
+
+  void UpdateMax(double new_max) final {
+    for (unsigned int i = 0; i < constraints.size(); i++) {
+      constraints[i].UpdateMax(new_max, false);
+    }
+  }
+
+  bool UpdateMinAndReturnBoolIfChanged(double new_min) final {
+    for (unsigned int i = 0; i < constraints.size(); i++) {
+      constraints[i].UpdateMin(new_min, true);
+    }
+    // even if nothing changed, this could have been unconstrained so it needs
+    // to be recomputed from the beginning
+    return true;
+  }
+
+  bool UpdateMaxAndReturnBoolIfChanged(double new_max) final {
+    for (unsigned int i = 0; i < constraints.size(); i++) {
+      constraints[i].UpdateMax(new_max, true);
+    }
+    // even if nothing changed, this could have been unconstrained so it needs
+    // to be recomputed from the beginning
+    return true;
+  }
+
+  FeatureConstraint *GetFeatureConstraint(int feature_index) final {
+    return &constraints[feature_index];
+  }
+};
+
 class BasicLeafConstraints : public LeafConstraintsBase {
  public:
   explicit BasicLeafConstraints(int num_leaves) : num_leaves_(num_leaves) {
-    entries_.resize(num_leaves_);
+    for (int i = 0; i < num_leaves; i++) {
```

Comment:
We always try to use prefix increments.
```suggestion
    for (int i = 0; i < num_leaves; ++i) {
```